### ISSUE TITLE
CA1859 not firing in System.Net.Mail source code

### ISSUE THREAD

**Issue by rzikm** [dotnet/roslyn-analyzers#7078]
### Analyzer

**Diagnostic ID**: [CA1859](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1859): `Use concrete types when possible for improved performance`

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: 7.0.0-preview1.22559.1

### Describe the bug

See https://github.com/dotnet/runtime/pull/95601#discussion_r1414422030

### Steps To Reproduce
```cs
public class C
{
    internal void ParseValue(string addresses)
    {
        IList<MailAddress> result = MailAddressParser.ParseMultipleAddresses(addresses); // should be flagged by CA1859

        for (int i = 0; i < result.Count; i++)
        {
            DoSomething(result[i]);
        }
    }
}
internal static class MailAddressParser
{
    internal static List<MailAddress> ParseMultipleAddresses(string data)
    {
        List<MailAddress> results = new List<MailAddress>();
        ... // Parse the string
        return results;
    }
}
```
### Expected behavior

Build fails (Treat warnings as errors is true) because of CA1859 on the following line

https://github.com/dotnet/runtime/blob/d7375b05ff2588884a4d2f4374d41cdbb9f07d27/src/libraries/System.Net.Mail/src/System/Net/Mail/MailAddressCollection.cs#L42

### Actual behavior

CA1859 is not detected, build passes.

CC @geeknoid

**Comment by geeknoid** [comment]
The issue is likely that the analyzer doesn't treat the indexer call as an interface call in the first place (where the affected code is doing `result[i] = xxx`). CA1859 only fires if the analyzer thinks it would eliminate some virtual or interface calls in the code.
